homeassistant/components/viam/__init__.py

```diff
@@ -0,0 +1,330 @@
+        confidence_threshold = float(call.data.get("confidence_threshold", 0.6))
+
+        async with await self._get_robot_client(
+            call.data.get("robot_secret"), call.data.get("robot_address")
+        ) as robot:
+            classifier = VisionClient.from_robot(robot, classifier_name)
+            image = Image.open(filepath) if filepath is not None else None
+            cam_bytes = (
+                await self._get_image_from_camera(camera) if camera is not None else b""
+            )
+            image = RawImage(cam_bytes, "jpeg") if len(cam_bytes) > 0 else None
+
+        if image is None:
+            return {
+                "classifications": [],
+                "img_src": filepath or None,
+            }
+
+        img_src = filepath or self._encode_image(image)
+        classifications = await classifier.get_classifications(image, count)
+
+        return {
+            "classifications": [
+                {"name": c.class_name, "confidence": c.confidence}
+                for c in classifications
+                if c.confidence >= confidence_threshold
+            ],
+            "img_src": img_src,
+        }
+
+    async def get_detections(self, call: ServiceCall) -> ServiceResponse:
+        """Accept input configuration to request detections."""
+        filepath = call.data.get("filepath")
+        camera = call.data.get("camera")
+        detector_name = call.data.get("detector_name")
+        confidence_threshold = float(call.data.get("confidence_threshold", 0.6))
+
+        async with await self._get_robot_client(
+            call.data.get("robot_secret"), call.data.get("robot_address")
+        ) as robot:
+            detector = VisionClient.from_robot(robot, detector_name)
+            image = Image.open(filepath) if filepath is not None else None
+            cam_bytes = (
+                await self._get_image_from_camera(camera) if camera is not None else b""
+            )
+            image = RawImage(cam_bytes, "jpeg") if len(cam_bytes) > 0 else None
+
+        if image is None:
+            return {
+                "detections": [],
+                "img_src": filepath or None,
+            }
+
+        img_src = filepath or self._encode_image(image)
+        detections = await detector.get_detections(image)
+
+        return {
+            "detections": [
+                {
+                    "name": c.class_name,
+                    "confidence": c.confidence,
+                    "x_min": c.x_min,
+                    "y_min": c.y_min,
+                    "x_max": c.x_max,
+                    "y_max": c.y_max,
+                }
+                for c in detections
+                if c.confidence >= confidence_threshold
+            ],
+            "img_src": img_src,
+        }
+
+    def _encode_image(self, image: RawImage):
+        image_string = (
+            str(base64.b64encode(image.data))
+            .replace("\\n", "")
+            .replace("b'", "")
```

Comment:
What are you doing here? `base64.b64encode` returns bytes, and this looks like trying to convert it to a string in a hacky way. Please search online for the proper way.

Reply:
This is creating an embeddable version of the image to be used in the `img_src` return value of the detection and classification services rather than depending on filesystem paths.

Reply:
`replace` is the fastest way to remove characters from a string ([reference](https://stackoverflow.com/a/27086669)) and the characters being removed in this method are invalid for displaying base64-encoded images.

Reply:
Can you use `image_string = base64.b64encode(image_bytes).decode()`(https://docs.python.org/3/library/stdtypes.html#bytes.decode) as you still try to convert bytes to a string.

Reply:
It looks like that work! Thanks for pointing out that method.